lib/fusion_jwt_authentication/jwks_strategy.ex

```diff
@@ -0,0 +1,131 @@
+  Refetch the jwks once each time a kid isn't found inside the jwks.
+  The jwks is cached in an ets table
+  """
+
+  require Logger
+
+  alias __MODULE__.EtsCache
+  alias Joken.Signer
+  alias FusionJWTAuthentication.API.JWT
+
+  def start_link(arg) do
+    GenServer.start_link(__MODULE__, arg, name: __MODULE__)
+  end
+
+  @doc false
+  def init(_) do
+    {:ok, %{}, {:continue, :do_init}}
+  end
+
+  def handle_continue(:do_init, _state) do
+    EtsCache.new()
+    fetch_signers()
+    {:noreply, %{}}
+  end
+
+  def match_signer_for_kid(kid) do
+    with {:cache, [{:signers, signers}]} <- {:cache, EtsCache.get_signers()},
+         {:signer, signer} when not is_nil(signer) <- {:signer, signers[kid]} do
+      {:ok, signer}
+    else
+      {:signer, nil} ->
+        fetch_signer_and_validate_kid(kid)
+
+      {:cache, []} ->
+        fetch_signer_and_validate_kid(kid)
+
+      err ->
+        err
+    end
+  end
+
+  def fetch_signer_and_validate_kid(kid) do
+    with true <- fetch_signers(),
+         {:cache, [{:signers, signers}]} <- {:cache, EtsCache.get_signers()},
+         {:signer, signer} when not is_nil(signer) <- {:signer, signers[kid]} do
+      {:ok, signer}
+    else
+      _ -> {:error, :kid_does_not_match}
+    end
+  end
+
+  @doc "Fetch signers"
+  def fetch_signers() do
+    with {:ok, {_status_code, %{"keys" => keys}}} <- JWT.jwks(),
+         {:ok, signers} <- validate_and_parse_keys(keys) do
+      Logger.debug("Fetched signers. #{inspect(signers)}")
+      EtsCache.put_signers(signers)
+    else
+      {:error, _reason} = err ->
+        Logger.warn("#{__MODULE__} failed to fetch signers. Reason: #{inspect(err)}")
+
+      err ->
+        Logger.warn("#{__MODULE__} got an unexpected error while fetching signers. Reason: #{inspect(err)}")
+    end
+  end
+
+  defp validate_and_parse_keys(keys) when is_list(keys) do
+    Enum.reduce_while(keys, {:ok, %{}}, fn key, {:ok, acc} ->
+      case parse_signer(key) do
+        {:ok, signer} -> {:cont, {:ok, Map.put(acc, key["kid"], signer)}}
+        e -> {:halt, e}
+      end
+    end)
+  end
+
+  defp parse_signer(key) do
+    with {:kid, kid} when is_binary(kid) <- {:kid, key["kid"]},
+         {:ok, alg} <- get_algorithm(key["alg"]),
+         {:ok, _signer} = res <- {:ok, Signer.create(alg, key)} do
+      res
+    else
+      {:kid, _} -> {:error, :kid_not_binary}
+      err -> err
+    end
+  rescue
+    e ->
+      Logger.warn("""
+      Error while parsing a key entry fetched from the network.
+
+      This should be investigated by a human.
+
+      Key: #{inspect(key)}
+
+      Error: #{inspect(e)}
+      """)
+
+      {:error, :invalid_key_params}
+  end
+
+  defp get_algorithm(nil), do: {:error, :no_algorithm_supplied}
+  defp get_algorithm(alg) when is_binary(alg), do: {:ok, alg}
+  defp get_algorithm(_), do: {:error, :bad_algorithm}
+end
+
+defmodule FusionJWTAuthentication.JWKS_Strategy.EtsCache do
+  @doc "Starts ETS cache"
+  def new do
```

Comment:
Why a use a public ets table. You do put yourself open for race conditions. 
Also you probably only need read_concurrency because 95% of the requests will read to get info from the current set of keys.

Would probably be better to still use handle_call for fetching new one from fusionauth. Because then you limit it to only 1 request to fusionauth. You can even put in the GenServer last lookup to ignore the request from fetching again.

Also it would be best to order them before you put it into ets if you combine that with an ordered_set ets table you can do quick lookups with :ets.lookup. (%{key => <jwks_token_data>}

Because now you are moving 20mb of key data from the ets table to the genserver and back and that will cause some issues.

Reply:
Just to explain a bit further. It's not bad to limit read requests to 1 at a time when key is not available.
Most of the requests should be fast because the keys are cached in the ets table.
So GenServer would only spend a small amount of time in a handle_call of the amount of requests from the user.
But will protect the service against duplicate requests to fusionauth when many users request at the same time and it's not available.
Also it would allow to put like a 5 sec timeout between requests if it's busy or something goes wrong.

Only reason it currently gives so much problems is that I underestimated the amount of different apps that need to load this data at the same time. And how slow that other endpoint was.